### PR TITLE
Feature/json api query

### PR DESCRIPTION
**Pull request by mcuyar**
Testing: 

run `docker-compose up`
run `docker exec -it luminary_api_1 phpunit`

Unit tests are located in the 'tests/ApiQuery' folder

View the Readme for documentation on how the API query handles requests

https://github.com/EngagePHP/Luminary/blob/c196f55f4ca2c4b77da26ac93a69e859d8b616c6/luminary/Services/ApiQuery/readme.md

### CODE REVIEW

File: luminary/Services/ApiQuery/Query.php [hunk 1]
```diff
@@ -0,0 +1,182 @@
+<?php
+
+namespace Luminary\Services\ApiQuery;
+
+use Illuminate\Database\Eloquent\Model;
+
+class Query
+{
+
```
Comment (bluewebtech) [inline luminary/Services/ApiQuery/Query.php]: Remove blank line.

File: luminary/Services/ApiQuery/QueryArr.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+     * Return an indexed array of dotted values to a multidimensional array
+     * with the last dot being the value
+     *
+     * @param  array $array
+     * @param string $defaultKey
+     * @return array
+     */
+    public static function dotValue(array $array, string $defaultKey = 'default') :array
+    {
+        $results = array();
```
Comment (bluewebtech) [inline luminary/Services/ApiQuery/QueryArr.php]: Convert old school array function to new [] fancy way.

File: luminary/Services/ApiQuery/QueryArr.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            $result = array();
+
+            switch (true) {
+                case $count <= 1:
+                    $result[$last_key] = [$value];
+                    break;
+                case $count > 1:
+                    $last = [$last_key => $value];
+                    foreach (array_reverse($item) as $key) {
+                        $temp = array();
```
Comment (bluewebtech) [inline luminary/Services/ApiQuery/QueryArr.php]: Convert old school array function to new [] fancy way.

File: luminary/Services/ApiQuery/QueryArr.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    /**
+     * Return a dotted array to a multi-dimensional associative array.
+     *
+     * @param  array   $array
+     * @param  string  $prepend
+     * @return array
+     */
+    public static function dotReverse($array, $prepend = '') :array
+    {
+        $results = array();
```
Comment (bluewebtech) [inline luminary/Services/ApiQuery/QueryArr.php]: Convert old school array function to new [] fancy way.

File: luminary/Services/ApiQuery/QueryArr.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+        foreach ($array as $context => $value) {
+            $context = explode('.', $context);
+            $count = count($context);
+            $last_key = array_pop($context);
+            $result = array();
```
Comment (bluewebtech) [inline luminary/Services/ApiQuery/QueryArr.php]: Convert old school array function to new [] fancy way.

File: luminary/Services/ApiQuery/QueryArr.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+            switch (true) {
+                case $count === 1:
+                    $result[$last_key] = $value;
+                    break;
+                case $count > 1:
+                    $last = [str_replace($prepend, '', $last_key) => $value];
+                    foreach (array_reverse($context) as $key) {
+                        $temp = array();
```
Comment (bluewebtech) [inline luminary/Services/ApiQuery/QueryArr.php]: Convert old school array function to new [] fancy way.

File: luminary/Services/ApiQuery/ServiceProvider.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Luminary\Services\ApiQuery;
+
+use Illuminate\Support\ServiceProvider as LaravelServiceProvider;
+use Illuminate\Database\Eloquent\Factory as ModelFactory;
+
+class ServiceProvider extends LaravelServiceProvider
+{
+
```
Comment (bluewebtech) [inline luminary/Services/ApiQuery/ServiceProvider.php]: Remove blank line.

File: luminary/Services/ApiQuery/ServiceProvider.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    /**
+     * Register the testing database migrations
+     * and factories
+     *
+     * @return void
+     */
+    public function registerTesting()
+    {
+        $dir = __DIR__.'/Testing/database';
```
Comment (bluewebtech) [inline luminary/Services/ApiQuery/ServiceProvider.php]: Do we format concatenated strings with spaces or no spaces? If spaces, anyway to add the requirement to the pre-commit config?

File: luminary/Services/ApiQuery/readme.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
+#JSON Api Query
```
Comment (bluewebtech) [inline luminary/Services/ApiQuery/readme.md]: Missing space between the # and the string.

File: luminary/Services/Filesystem/App/Storage.php [hunk 10]
```diff
@@ -22,4 +23,67 @@ protected static function getStorage()
...
+    /**
+     * Create a relative link to the target file or directory.
+     *
+     * @param  string  $target
+     * @param  string  $link
+     * @return string
+     */
+    public static function link($target, $link) :string
+    {
+        if (! is_link($target)) {
```
Comment (bluewebtech) [inline luminary/Services/Filesystem/App/Storage.php:74]: Saw this in a few other spots. Do we add space after logical operator or is this an oopsi?

Comment (mcuyar) [reply luminary/Services/Filesystem/App/Storage.php:74]: I always do, but I guess the answer is that I have no clue

File: luminary/Services/FormRequests/FormRequest.php [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+     * @param  array  $errors
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    public function response(array $errors)
+    {
+        if ($this->expectsJson()) {
+            return new JsonResponse($errors, 422);
+        }
+
+        return $this->redirector->to($this->getRedirectUrl())
```
Comment (bluewebtech) [inline luminary/Services/FormRequests/FormRequest.php:144]: Formatting is a bit funky here.

File: luminary/Services/ApiQuery/ServiceProvider.php (hunk 8 above)
Comment (mcuyar) [reply luminary/Services/ApiQuery/ServiceProvider.php]: We can write our own rules. Been wanting to do that for doc blocks, etc. Maybe we can add it next sprint?

File: luminary/Services/Generators/Creators/Console/Kernel.php [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        $commands = array_map(function ($command) {
+            $command = trim($command);
+            $command = preg_replace('/(^[\"\'\s]|[\"\'\s]$)/', '', $command);
+            $command = preg_match("/::class/i", $command) ? $command : '\''.$command.'\'';
+
+            return "        " . trim($command);
+        }, $commands);
+
+        // Create the final formatted output
+        $commands = PHP_EOL . implode(','.PHP_EOL, $commands) . PHP_EOL . "    ";
```
Comment (bluewebtech) [inline luminary/Services/Generators/Creators/Console/Kernel.php]: Add space to concat section - ','.PHP_EOL

File: luminary/Services/Generators/Creators/Tests/PhpUnit.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace Luminary\Services\Generators\Creators\Tests;
+
+use Luminary\Services\Generators\Creators\StubCreator;
+
+class PhpUnit extends StubCreator
+{
+
```
Comment (bluewebtech) [inline luminary/Services/Generators/Creators/Tests/PhpUnit.php]: Remove blank line.

**Review by bluewebtech — APPROVED** [review]
Few small things that seem common.

Do we add space between concat strings? 
Do we add space after logical operator within an if block?

File: luminary/Services/FormRequests/FormRequest.php (hunk 11 above)
Comment (mcuyar) [reply luminary/Services/FormRequests/FormRequest.php:144]: I fixed this in the last push

**Pull request merged**